doc/unboxing.md

```diff
@@ -0,0 +1,234 @@
+Unboxing
+========
+
+## Overview
+
+The unboxing algorithm in Flambda 2 aims to remove unnecessary allocations
+without causing bad behaviour (e.g. pushing allocations into loops).
+It currently supports:
+
+- Unboxing along non-looping control flow, within a single function, of the
+  following varieties of values:
+  - Boxed `float`
+  - Boxed `int32`
+  - Boxed `int64`
+  - Boxed `nativeint`
+  - Variants
+  - Closures (need examples).
```

Comment:
Records (i.e. fixed-size blocks) are also unboxed, and although they may be considered as a corner case of variants, I think that records are sufficiently distinct from variants in the source language and in practice to mention them explicitly here.